### ISSUE TITLE
General support ticket for Qt 6

### ISSUE THREAD

**Issue by mossheim** [supercollider/supercollider#5169]
This is a ticket for general discussion on Qt 6.

Qt 6 will require c++17, which we are tracking on another ticket.

-----
Edited by @dyfer to help with keeping track of the progress; last update 2024-11-04.
Original TODO from @mossheim 
- [x] fix register in SCDoc/lex.scdoc.cpp #5396
- [x] upgrade to C++17 #5396
- [x] fix Qt deprecation warnings, ~~keeping compatibility with Qt 5.9 if possible~~ #6464
- [x] determine whether to deprecate QuartzComposer code - deprecated in #5710
- [x] migrate any remaining code incompatible with Qt6 #6475
- [x] determine what code relies on Core5Compat and migrate as much as possible #6475 (except regex code)
- [x] port search/replace regex code https://github.com/supercollider/supercollider/pull/6512
- [x] remove Core5Compat from cmake and build system (after all code relying on it is migrated) ##6524

Use the ~~[Qt6 project](https://github.com/supercollider/supercollider/projects/21)~~ [Qt6 milestone](https://github.com/supercollider/supercollider/milestone/26) to keep track of open PRs.

**Comment by brianlheim** [comment]
here is a branch which at least compiles with the Qt 6.0.0-beta2 release: https://github.com/brianlheim/supercollider/tree/topic/qt6

some notes i have:
- Qt6 provides a Core5Compat module which saves us from having to migrate too many things at once
- seems like the MacExtras module is gone?
- many compile errors / commented out code from my branch would likely be fixed simply by addressing deprecation warnings under Qt 5.15. that should be the first priority.
- it doesn't seem like QuartzComposer/cocoa stuff is present in qt6, at least i couldn't find it. if it's truly gone, i think we should plan to deprecate it in 3.12 to give people a heads up
- not sure the SafePtr replacement is correct but worth looking at in more detail. no reason for us to reimplement that now.
- since this requires C++17, the `register`s in SCDoc/lex.scdoc.cpp have to go. there should probably be an option in lex to not generate them, i just removed them without much concern for that in order to compile.
- need to find what the correct replacement for the cmake command qt6_add_translation is
- the only thing which seems to require some real thought is moving from QRegExp to QRegularExpression in autocompleter.cpp, since we must interop with a Qt object that requires it
- supporting both qt5 and qt6 is probably doable for awhile

overall this is a *lot* less painful than i was expecting, kudos to Qt for that!

**Comment by brianlheim** [comment]
here's a rough roadmap for upgrading:
- fix `register` in SCDoc/lex.scdoc.cpp
- upgrade to C++17
- fix Qt deprecation warnings, keeping compatibility with Qt 5.9 if possible
- port autocompleter regex code
- determine whether to deprecate QuartzComposer code
- migrate any remaining code incompatible with Qt6
- determine what code relies on Core5Compat and migrate as much as possible

**Comment by dyfer** [comment]
Just so this is written down somewhere: I've been making an on and off progress on this, based in @jrsurge 's work. The latest branch is here: https://github.com/supercollider/supercollider/tree/topic/qt6-07 (you can find some earlier attempts too under `topic/qt6-0*`)

In this branch SC compiles with Qt6 on macOS, but it crashes when attempting to open a Help Browser window. I have no confidence in changes I made in https://github.com/supercollider/supercollider/commit/a31d66728f4c565c969ac58ab729f75a48847e0b; as a side effect the [change in QcCallback](https://github.com/supercollider/supercollider/blob/topic/qt6-07/QtCollider/QcCallback.hpp#L25) broke building with `SC_USE_QTWEBENGINE=OFF` flag (otherwise I got SC build with qt6 without the webengine a while back, before this latest attempt at making the webengine work; e.g [this branch](https://github.com/supercollider/supercollider/tree/topic/qt6-02-webengine) used to build without the webengine ).

**Comment by dyfer** [comment]
The runtime crash seems to be happening due to issue with deployment with macdeployqt on macOS with Qt from homebrew. The build against the official qt binaries seems to be working fine. This is the branch that seems to be building fine: https://github.com/supercollider/supercollider/tree/topic/qt6-09 (there is a minor cmake change in it, in comparison with qt6-08). 
There are still some build issues on Linux and Windows that need to be fixed...

**Comment by dyfer** [comment]
The branch https://github.com/supercollider/supercollider/tree/topic/qt6-12 works with qt6 _in principle_. It is not PR-ready yet, here's what's missing:
- local macOS builds with qt6 from homebrew have improperly assembled app bundles and crash when trying to run. This seems to be a problem with `macdeployqt` handling of paths that's between [homebrew](https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue+macdeployqt+is%3Aclosed) and [qt](https://bugreports.qt.io/browse/QTBUG-84200?jql=text%20~%20%22macdeployqt%20homebrew%22%20ORDER%20BY%20affectedVersion%20DESC).. I'm not sure what's the path forward in this case.
- Windows build works, but I'd like the UTF-16 changes to be reviewed and merged first, then I'll rebase the qt6 branch on develop https://github.com/supercollider/supercollider/pull/6124
- Linux builds with clang fail to link with Qt6. I'm asking for help in that regard in #6126

**Comment by dyfer** [comment]
Update:
https://github.com/supercollider/supercollider/tree/topic/qt6-21 is my most current branch with the fixes.
- I fixed (patched) deployment with macdeplyqt from homebrew
- #6124 is waiting for review: #6126 is merged
- current work: I realized that I'm sort of close to not having to use the qt5compat module. The main non-trivial change is replacing QRegExp with QRegularExpression. A WIP branch is [here](https://github.com/dyfer/supercollider/commits/topic/QRegularExpression-02), however it is not working properly yet.

My plan is to finish QRegularExpression migration and submit it as a separate PR (these changes are compatible with Qt5). Once that PR is ready, reviewed and merged, and once #6124, I'll rebase the `qt6-xx` branch on develop and submit a PR.

**Comment by dyfer** [comment]
A quick update: I have not resumed work on the port.
The most up-to-date branch with my changes is here: https://github.com/dyfer/supercollider/tree/topic/qt6-21
Adding to the previous comments and timeline, I later realized that my porting of the QtCallback mechanism does not actually work (it does compile at least). So the changes in https://github.com/dyfer/supercollider/commit/a9470a4d4aa3b6aaabd7ceaa315a70b835add629 need to be reviewed and fixed...

**Comment by xunil-cloud** [comment]
Hi, I actually looked into the [dyfer@a9470a4](https://github.com/dyfer/supercollider/commit/a9470a4d4aa3b6aaabd7ceaa315a70b835add629) commit and finally figured out what was wrong with it!

Since the actual type information is stored in `type`, so  we can't use `QVariant::fromValue` here. Otherwise, we always get a void pointer from `QVariant`. (the type of `argData[i + 1]`  is `void *`)
```diff
diff --git a/QtCollider/QcSignalSpy.h b/QtCollider/QcSignalSpy.h
index 496ca244fd..38b3d8933c 100644
--- a/QtCollider/QcSignalSpy.h
+++ b/QtCollider/QcSignalSpy.h
@@ -97,13 +97,14 @@ public:
                     // avoid creating a QVariant<QVariant>
                     args << QVariant(type, argData[i + 1]).value<QVariant>();
 #else
-                    args << QVariant::fromValue(argData[i + 1]).value<QVariant>();
+                    args << QVariant(QMetaType(type), argData[i + 1]).value<QVariant>();
 #endif
                 } else {
 #if (QT_VERSION < QT_VERSION_CHECK(6, 0, 0))
                     args << QVariant(type, argData[i + 1]);
 #else
-                    args << QVariant::fromValue(argData[i + 1]);
+
+                    args << QVariant(QMetaType(type), argData[i + 0]);
 #endif
                 }
             }
```

And here, changing `[&cb]` to `[cb]` seems to work
```diff
diff --git a/QtCollider/widgets/QcWebView.cpp b/QtCollider/widgets/QcWebView.cpp
index 9b00250e8c..69716ef40d 100644
--- a/QtCollider/widgets/QcWebView.cpp
+++ b/QtCollider/widgets/QcWebView.cpp
@@ -203,11 +203,9 @@ void WebView::findText(const QString& searchText, bool reversed, QcCallback* cb)
     if (!cb) {
         QWebEngineView::findText(searchText, flags);
     } else {
-        QWebEngineView::findText(searchText, flags, cb->asFunctor());
-        // QWebEngineView::findText(searchText, flags, [&cb](const QWebEngineFindTextResult& result) {
-        //     std::cout << result.activeMatch() << "of" << result.numberOfMatches() << "matches" << std::endl;
-        //     cb->asFunctor()(result.numberOfMatches() > 0);
-        // });
+        QWebEngineView::findText(searchText, flags, [cb](const QWebEngineFindTextResult& result) {
+            cb->asFunctor()(result.numberOfMatches() > 0);
+        });
     }
 }

diff --git a/QtCollider/widgets/QcWebView.h b/QtCollider/widgets/QcWebView.h
index 403a56cac9..cd53f50290 100644
--- a/QtCollider/widgets/QcWebView.h
+++ b/QtCollider/widgets/QcWebView.h
@@ -67,7 +67,6 @@ Q_SIGNALS:
     void onCalled(const QString&);
     void onCalled(const QVariant&);
     void onCalled(const QUrl&);
-    void onCalled(const QWebEngineFindTextResult&);
 };
```

**Comment by xunil-cloud** [comment]
Oh, `QcSignalSpy.h` file is actually changed in this [commit](https://github.com/dyfer/supercollider/commit/4e90af579635cebb30f057b568f6e91e19aa186a).

**Comment by dyfer** [comment]
> And here, changing `[&cb]` to `[cb]` seems to work

Thanks @xunil-cloud ! That doesn't seem to be all what's needed - it would compile, but finding text in the web page didn't work.
Recently I got another go at this... Didn't get much further, but I think that using `QWebEngineFindTextResult` is not the right track, as it provides only information about search (which result is active), as opposed to the actual search result string, i.e. the `findText` signal changed in Qt6 and doesn't provide the actual string result anymore - but the `findTextFinished` signal does. 

I now reverted moving the QcCallback from what I had in the qt-21 branch. I made a new branch [topic/qt-26-cleanWebView](https://github.com/dyfer/supercollider/tree/topic/qt-26-cleanWebView) where I removed the changes to QcCallback, as I don't think they are needed (I might be wrong). 

**I think that branch is a good starting** point to figuring out how to address the findText issue. I _think_ the changes need to happen in QcWebView.hpp by adding a signal for `findTextFinised` somewhere there ([my failed attempt here](https://github.com/dyfer/supercollider/commit/c32a8c27ecf1757f03ac5ca5309a539ab5bc40b8#diff-1d69f5603af53611d078b7cfd6293cad5f8c2333ab2fcc3db248be8fac5db46eR195-R207)), but I'm not sure if that should be done inside the findText function, or added to the signals in the `connectPage` method.

BTW the `topic/qt-26-cleanWebView` branch compiles in qt5 but not in qt6.

@xunil-cloud if you are willing to take a look at this, I would really appreciate, thanks!

**Comment by xunil-cloud** [comment]
That's weird. I believe `findText` should works now. To get everything working, you'll also need to update the `QcSignalSpy.h` file, otherwise, sclang interpreter can't handle Qt signal properly. I push my change on  [qt6-webview](https://github.com/xunil-cloud/supercollider/commits/qt6-webview/) branch based on [topic/qt-26-cleanWebView](https://github.com/dyfer/supercollider/tree/topic/qt-26-cleanWebView).

How did you test the findText function? I use little sclang script to test on my Linux system with Qt 6.7.2. The changes also pass TestWebView test.

**Comment by dyfer** [comment]
Ah thanks so much @xunil-cloud ! Your changes indeed worked!
I'll now prepare the proper PR to migrate to Qt6, incorporating your changes (pulled into [this branch](https://github.com/dyfer/supercollider/tree/topic/qt6-27), I hope that's okay?).

Again, thanks for pointing out the missing elements.

**Comment by xunil-cloud** [comment]
>I hope that's okay?

Yep. that's okay :smile: 

>I tested findText with the code taken from the TestWebView unit test

OK. I do something similar, but with modifications based on the doc example.

**Comment by dyfer** [comment]
@xunil-cloud thanks again!

To update everyone on this:
Qt6 port seems to be working now (minor nuances to be ironed out in the review process notwithstanding). 
Now I think we have a path forward. Once #6331 is pulled in (since it touches GHA build matrix) I'll prepare the Qt6 update in 3 PRs for easier review: 
1. All "trivial" deprecation changes, updating Qt requirement to Qt 5.15
2. Qt6 changes, including CMake and GHA support (after 1. is reviewed and merged)
3. Documentation changes, updating information about Qt version, build instructions etc (concurrently with 2.)

Afterwards the only leftover from the todo list in this issue will be porting regex code, after which I think we won't need `Core5Compat` module anymore. I'm not sure if that should become a separate issue, or should we keep this one open until the regex code is ported....

Let me know if you have questions and comments!

**Comment by sonata-chen** [comment]
Hi, I’m continuing @dyfer’s work on [QRegularExpression](https://github.com/sonata-chen/supercollider/commits/topic/QRegularExpression/). Syntax highlighting and Find and Replace should be working now. You can try it out!

**Comment by sonata-chen** [comment]
yes, I plan to do that as well.

**Comment by dyfer** [comment]
Hello @sonata-chen, I wanted to check in - are you still considering making a PR for QRegularExpression? Thanks!

**Comment by sonata-chen** [comment]
> are you still considering making a PR for QRegularExpression? Thanks!

Yes, i have already rebased my branch onto `develop` branch. A PR will be opened later.

**Comment by dyfer** [comment]
I think we can close this now - the outlined plan for migrating to Qt6 has been completed. 🎉

Thanks everyone who helped along the way! @mossheim, @jrsurge, @capital-G, @xunil-cloud, @sonata-chen, @JordanHendersonMusic

Remaining issues pertaining to qt6 are tracked in the **[Qt6 milestone](https://github.com/supercollider/supercollider/milestone/26)**.
